src/charm.py

```diff
@@ -68,16 +70,37 @@ def _on_legacy_relation_created(self, event: ops.RelationCreatedEvent) -> None:
     def _on_config_changed(self, _) -> None:
         """Handle changes in configuration."""
         self.unit.status = ops.MaintenanceStatus("Configuring charm")
-        self._store_password_as_secret()
+        if self._has_secrets():
+            self._store_password_as_secret()
         self._update_relations()
         self.unit.status = ops.ActiveStatus()
 
-    def _store_password_as_secret(self) -> None:
-        """Store the SMTP password as a secret."""
-        if self._charm_state.password and self._has_secrets():
-            # Note https://github.com/juju/python-libjuju/issues/970
+    def _store_password_as_secret(self) -> Optional[ops.Secret]:
+        """Store the SMTP password as a secret.
+
+        Returns:
+            the secret id.
+        """
+        peer_relation = self.model.get_relation("smtp-peers")
+        assert peer_relation  # nosec
+        secret = None
+        if secret_id := peer_relation.data[self.app].get("secret-id"):
```

Comment:
I'm wondering if this could be in `charm_state.py`?

Reply:
I've considered it but finally decided to leave it here since this is a juju value derived from config. What are your thoughts?

Reply:
I guess it does make sense to have a more cohesive code for the function that passes data on as secret, thanks!